Make sure to reload subresources loaded during the load of a hard reloaded document

youennf opened the pull request:
#### 3612aa5748ea37b492f8b45541eaaae130877779
<pre>
Make sure to reload subresources loaded during the load of a hard reloaded document
<a href="https://bugs.webkit.org/show_bug.cgi?id=250705">https://bugs.webkit.org/show_bug.cgi?id=250705</a>
rdar://36246099

Reviewed by Alex Christensen.

When doing a page reload from origin, we should ensure to reload from origin the subresources loaded during the initial load.
This is on par with Chrome and Firefox.
To do so, we set the fetch cache to reload in case page is being reloaded from origin.

Add a test that checks through service worker that we set the cache mode of a request accordingly.
Update inspector tests that make use of reload from origin.

* LayoutTests/http/tests/inspector/network/har/har-page-aggressive-gc-expected.txt:
* LayoutTests/http/tests/inspector/network/har/har-page-expected.txt:
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker.html: Added.
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker.js: Added.
(async doClaim):
(onactivate):
(onfetch.async e):
(onmessage):
* LayoutTests/http/wpt/service-workers/cache-test-serviceworker.html: Added.
* LayoutTests/http/wpt/service-workers/cache-test-serviceworker.js: Added.
(async doClaim):
(onactivate):
(onfetch.async e):
* LayoutTests/http/wpt/service-workers/resources/cache-mode-hard-reload-serviceworker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/resources/test.py: Added.
(main):
* LayoutTests/platform/mac-wk1/http/tests/inspector/network/har/har-page-expected.txt:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateHTTPRequestHeaders):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::updateCacheModeIfNeeded):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::reloadFromOrigin):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::reloadFromOrigin):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::reloadFromOrigin):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/259019@main">https://commits.webkit.org/259019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/913092abe28c12ba66f8b6b0d8e5ea260f9ed88b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112787 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172993 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107504 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3566 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95808 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111959 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109324 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93613 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38279 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92365 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25209 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79920 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6046 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26611 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6226 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3138 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12208 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46121 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6194 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7979 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -891,6 +891,8 @@ void CachedResourceLoader::updateHTTPRequestHeaders(FrameLoader& frameLoader, Ca
 #endif // ENABLE(PUBLIC_SUFFIX_LIST)
     request.updateUserAgentHeader(frameLoader);
 
+    if (frameLoader.frame().loader().loadType() == FrameLoadType::ReloadFromOrigin)
+        request.updateCacheModeIfNeeded(cachePolicy(type, request.resourceRequest().url()));
```
I wonder if we should also update the cache mode here if Web Inspector has turned on the "Disable cache" option? (e.g. `page->isResourceCachingDisabledByWebInspector()`).

youennf replied:
We could try to do that as a simplification and maybe remove ResourceRequestCachePolicy::ReloadIgnoringCacheData.

Web Inspector is using ResourceRequestCachePolicy::ReloadIgnoringCacheData to bypass the disk cache. With this approach, service workers do not see the cache is bypassed, which is probably ok for web inspector.
With this patch change, the service worker will see caches are bypassed when doing a hard reload, which matches other browsers behavior.

youennf committed:
Make sure to reload subresources loaded during the load of a hard reloaded document
https://bugs.webkit.org/show_bug.cgi?id=250705
rdar://36246099

Reviewed by Alex Christensen.

When doing a page reload from origin, we should ensure to reload from origin the subresources loaded during the initial load.
This is on par with Chrome and Firefox.
To do so, we set the fetch cache to reload in case page is being reloaded from origin.

Add a test that checks through service worker that we set the cache mode of a request accordingly.
Update inspector tests that make use of reload from origin.

* LayoutTests/http/tests/inspector/network/har/har-page-aggressive-gc-expected.txt:
* LayoutTests/http/tests/inspector/network/har/har-page-expected.txt:
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker.html: Added.
* LayoutTests/http/wpt/service-workers/cache-mode-hard-reload-serviceworker.js: Added.
(async doClaim):
(onactivate):
(onfetch.async e):
(onmessage):
* LayoutTests/http/wpt/service-workers/cache-test-serviceworker.html: Added.
* LayoutTests/http/wpt/service-workers/cache-test-serviceworker.js: Added.
(async doClaim):
(onactivate):
(onfetch.async e):
* LayoutTests/http/wpt/service-workers/resources/cache-mode-hard-reload-serviceworker.py: Added.
(main):
* LayoutTests/http/wpt/service-workers/resources/test.py: Added.
(main):
* LayoutTests/platform/mac-wk1/http/tests/inspector/network/har/har-page-expected.txt:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateHTTPRequestHeaders):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::updateCacheModeIfNeeded):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::reloadFromOrigin):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::reloadFromOrigin):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::reloadFromOrigin):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/259019@main
